### PR TITLE
Eliminate Useless Cast during Transformer.

### PR DESCRIPTION
**Pull request by iK1D**
Eliminate Cast nodes if the input's data type is same as the 'to' type of this node.

**Motivation and Context**
- For current UniLM v2 model implementation, this can eliminate Cast nodes from number of 70+ to around 40.

### CODE REVIEW

File: onnxruntime/core/optimizer/graph_transformer_utils.cc [hunk 1]
```diff
@@ -51,6 +52,7 @@ std::vector<std::unique_ptr<RewriteRule>> GenerateRewriteRules(TransformerLevel
       rules.push_back(onnxruntime::make_unique<ConvAddFusion>());
       rules.push_back(onnxruntime::make_unique<ConvMulFusion>());
       rules.push_back(onnxruntime::make_unique<ConvBNFusion>());
+      rules.push_back(onnxruntime::make_unique<CastElimination>());
```
Comment (SherlockNoMad) [inline onnxruntime/core/optimizer/graph_transformer_utils.cc:55]: let's follow the original pattern here
elimination first, fusion latter

**Comment by SherlockNoMad** [comment]
we should also add this to orttraining/core/optimzier/graph_transformer_utils.cc 

If some nodes are redundant, no need to build the gradient for them

File: onnxruntime/core/optimizer/cast_elimination.cc [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  return Status::OK();
+}
+
+bool CastElimination::SatisfyCondition(const Graph&, const Node& node, const logging::Logger&) const {
+  const auto* input_type = node.InputDefs()[0]->TypeAsProto();
+  if (input_type == nullptr || !input_type->tensor_type().has_elem_type()) {
+    return false;
+  }
+
+  return optimizer_utils::IsAttributeWithExpectedValue(node, "to", static_cast<int64_t>(input_type->tensor_type().elem_type()));
```
Comment (pengwa) [inline onnxruntime/core/optimizer/cast_elimination.cc:27]: should we add graph_utils::CanRemoveNode here?

File: onnxruntime/test/testdata/transform/cast_elimination.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+less2 = helper.make_node('Less', ['x1', 'x3'], ['less2'], name='less2')
+cast1 = helper.make_node('Cast', ['less1'], ['cast1'], name='cast1', to=9)
+and_node = helper.make_node('And', ['cast1', 'less2'], ['and_node'], name='and_node')
+cast2 = helper.make_node('Cast', ['and_node'], ['cast2'], name='cast2', to=9)
+cast3 = helper.make_node('Cast', ['cast2'], ['cast3'], name='cast3', to=1)
+cast4 = helper.make_node('Cast', ['x1'], ['cast4'], name='cast4', to=1)
+matmul = helper.make_node('MatMul', ['cast3', 'cast4'], ['matmul'], name='matmul')
+cast5 = helper.make_node('Cast', ['matmul'], ['cast5'], name='cast5', to=7)
+cast6 = helper.make_node('Cast', ['cast5'], ['cast6'], name='cast6', to=7)
+identity = helper.make_node('Identity', ['cast6'], ['output'], name='identity')
```
Comment (pengwa) [inline onnxruntime/test/testdata/transform/cast_elimination.py:22]: cast as graph outputs is also an important test case.

File: onnxruntime/core/optimizer/cast_elimination.cc (hunk 2 above)
Comment (iK1D) [reply onnxruntime/core/optimizer/cast_elimination.cc:31]: Added.

File: onnxruntime/test/testdata/transform/cast_elimination.py (hunk 3 above)
Comment (iK1D) [reply onnxruntime/test/testdata/transform/cast_elimination.py]: Added some more Cast nodes around input/output for test.

**Pull request merged**